### ISSUE TITLE
[SystemZ] constexpr variable promoted to global does not minimum alignment

### ISSUE THREAD

**Issue by nikic** [llvm/llvm-project#224008]
```cpp
char test(unsigned n) {
  constexpr char digits[] = "foobar";
  return digits[n];
}
```
`clang --target=s390x-unknown-linux-gnu -S -emit-llvm -o - -O2 test.cpp` produces:
```llvm
@__const._Z4testj.digits = private unnamed_addr constant [7 x i8] c"foobar\00", align 1
```
Because it does not have the minimum global alignment of 2, it gets lowered to this GOT reference:
```
	lgrl	%r1, .L__const._Z4testj.digits@GOT
```

This is probably not supposed to happen? I expect either Clang should emit this with align 2, or the backend should raise the alignment (it has private linkage), or both.

**Comment by nikic** [comment]
What I actually ran into is a miscompile where the GOT slot only has the low 32-bits of the address. I guess that's the failure behavior we get from https://github.com/llvm/llvm-project/issues/224011 in a build without assertions.

**Comment by nikic** [comment]
cc @uweigand

**Comment by uweigand** [comment]
I would have expected clang to use `align 2` here to respect the `MinGlobalAlign` setting.  However, it seems that while this is done in `getDeclAlign` for variables that were originally global, it is *not* respected for variables made global by `createUnnamedGlobalFrom
` - this reuses the alignment the variable had *before* it was made global, so `MinGlobalAlign` is not taken into account.

In the case where the backend does see the `align 1` case, it will indeed try to make it work by putting the unaligned address into the GOT and loading it from there.   This should actually work fine (if a bit slower), so if it doesn't, that seems to be an lld bug.  I'll have a look at the lld issue.

**Comment by nikic** [comment]
Fixed by https://github.com/llvm/llvm-project/pull/224586, raising the alignment in the backend rather than frontend.
